### ISSUE TITLE
Facilitate automatic form and client generation

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#6487]
A number of people are building UIs and/or text-based configuration layers and/or programmatic clients for Kubernetes.

It shouldn't be necessary to create manually maintained wrappers that mirror most fields of the Kubernetes API, especially if those fields are just going to be exposed pretty much directly to users. We have this problem ourselves (#4874).

What is needed beyond swagger-based code generation (and a correct+complete swagger description of the API)?
#4210 and #5889 requested better validation support.

We've discussed supporting object generators such as run-container as services rather than just in clients. #5280

What else?

cc @brendandburns @nikhiljindal @lavalamp @vishh @iantw89 @jackgr @simon3z @abonas @jstrachan @preillyme @gokrokvertskhov @craiglpeters @ghodss @smarterclayton @madhuri-rai07 @nirmal070125 @Ch00k

**Comment by smarterclayton** [comment]
@jwforres

----- Original Message -----

**Comment by jwforres** [comment]
In general being able to do client side validation leads to a snappier feeling GUI (dont have to wait on a server response to tell the user there was an error with their input), but to do that the client needs to know what it should be validating.  The basics go a long way, e.g. what fields are required, the expected type (string, bool, int, etc), min/max values (ints/floats), min/max length (strings).   Nice to have would be returning a regexp to validate against.

CC @csrwng @jcantrill

**Comment by nirmal070125** [comment]
Awesome idea to have Swagger support :-)

**Comment by smarterclayton** [comment]
For object references we need to do a better job of defining in swagger what the expectations are (for references that are multi kind, what allowed kinds are).

Label selectors need to be identified by a type.

I don't believe the swagger output for types is complete today (large sections appear missing [citation needed])

Described defaults for the UI to prefill / background

Fields that are not editable after creation.

Fields that should not automatically be included in clone/new operations (service portalIP, name).

Associated resources (implicit or otherwise), such as RC -> pods and service -> endpoints

Pod template section should be identified as a template type.

**Comment by bgrant0607** [comment]
@nirmal070125 We have swagger support, which we're improving. The master exports /swaggerapi, and we statically generate the description periodically and put it here:
https://github.com/GoogleCloudPlatform/kubernetes/tree/master/api/swagger-spec
and
kubernetes.io/third_party/swagger-ui/

Known swagger issues are #6457, #2971, #4700, #2968, #6167, #5331.

**Comment by nirmal070125** [comment]
Thanks for the pointers Brian!

**Comment by bgrant0607** [comment]
@smarterclayton Thanks for the list.

Known swagger issues are listed in my previous comment. Please file new issues for other problems.

Selector type is covered by #341. 

Links to related objects: #3676, #1348

Readonly (#2970) and not-updatable fields (#1728) have been documented in swagger, but not in a machine-readable way. Swagger 2.0 (we're on 1.2) supports readonly, but we may need an additional feature or convention to represent not-updatable.

**Comment by smarterclayton** [comment]
OpenShift added an "edit" command that uses your shell / system editor to edit a resource (open kubectl get in vi, save with kubectl update).  I'll port it to Kube once I have support for merge patches on conflicts, but it would benefit from annotation of fields from swagger like the items below.

----- Original Message -----

**Comment by bgrant0607** [comment]
For "edit" and for documentation generation, another thing that would be useful is identification of the most commonly used fields, together with example values, as proposed in https://github.com/GoogleCloudPlatform/kubernetes/issues/6487#issuecomment-90353795, so that users aren't overwhelmed.

**Comment by bgrant0607** [comment]
Port info: #6908

**Comment by kargakis** [comment]
@kubernetes/sig-api-machinery-misc this seems fixed now?

**Comment by madhuri-rai07** [comment]
FYI, [1] is swagger generated python client. But it hardly worked so we had to manually correct those issues and even we posted some fixes to swagger codegen.

[1] https://github.com/openstack/python-k8sclient/tree/master/k8sclient

**Comment by bgrant0607** [comment]
UI-oriented form generation layered on JSON schema: http://schemaform.io/
